src/code/z_message_PAL.c

```diff
@@ -1,66 +1,3306 @@
+
+    if (sIconPrimR >= sIconPrimColors[sIconFlashColorIdx].r) {
+        sIconPrimR -= primR;
+    } else {
+        sIconPrimR += primR;
+    }
+
+    if (sIconPrimG >= sIconPrimColors[sIconFlashColorIdx].g) {
+        sIconPrimG -= primG;
+    } else {
+        sIconPrimG += primG;
+    }
+
+    if (sIconPrimB >= sIconPrimColors[sIconFlashColorIdx].b) {
+        sIconPrimB -= primB;
+    } else {
+        sIconPrimB += primB;
+    }
+
+    envR = (ABS(sIconEnvR - sIconEnvColors[sIconFlashColorIdx].r)) / sIconFlashTimer;
+    envG = (ABS(sIconEnvG - sIconEnvColors[sIconFlashColorIdx].g)) / sIconFlashTimer;
+    envB = (ABS(sIconEnvB - sIconEnvColors[sIconFlashColorIdx].b)) / sIconFlashTimer;
+
+    if (sIconEnvR >= sIconEnvColors[sIconFlashColorIdx].r) {
+        sIconEnvR -= envR;
+    } else {
+        sIconEnvR += envR;
+    }
+
+    if (sIconEnvG >= sIconEnvColors[sIconFlashColorIdx].g) {
+        sIconEnvG -= envG;
+    } else {
+        sIconEnvG += envG;
+    }
+
+    if (sIconEnvB >= sIconEnvColors[sIconFlashColorIdx].b) {
+        sIconEnvB -= envB;
+    } else {
+        sIconEnvB += envB;
+    }
+
+    sIconFlashTimer--;
+
+    if (sIconFlashTimer == 0) {
+        sIconPrimR = sIconPrimColors[sIconFlashColorIdx].r;
+        sIconPrimG = sIconPrimColors[sIconFlashColorIdx].g;
+        sIconPrimB = sIconPrimColors[sIconFlashColorIdx].b;
+        sIconEnvR = sIconEnvColors[sIconFlashColorIdx].r;
+        sIconEnvG = sIconEnvColors[sIconFlashColorIdx].g;
+        sIconEnvB = sIconEnvColors[sIconFlashColorIdx].b;
+        sIconFlashTimer = 12;
+        sIconFlashColorIdx ^= 1;
+    }
+
+    gDPPipeSync(gfx++);
+
+    gDPSetCombineLERP(gfx++, PRIMITIVE, ENVIRONMENT, TEXEL0, ENVIRONMENT, TEXEL0, 0, PRIMITIVE, 0, PRIMITIVE,
+                      ENVIRONMENT, TEXEL0, ENVIRONMENT, TEXEL0, 0, PRIMITIVE, 0);
+
+    gDPSetPrimColor(gfx++, 0, 0, sIconPrimR, sIconPrimG, sIconPrimB, 255);
+    gDPSetEnvColor(gfx++, sIconEnvR, sIconEnvG, sIconEnvB, 255);
+
+    gDPLoadTextureBlock_4b(gfx++, iconTexture, G_IM_FMT_I, FONT_CHAR_TEX_WIDTH, FONT_CHAR_TEX_HEIGHT, 0,
+                           G_TX_NOMIRROR | G_TX_CLAMP, G_TX_NOMIRROR | G_TX_CLAMP, G_TX_NOMASK, G_TX_NOMASK, G_TX_NOLOD,
+                           G_TX_NOLOD);
+
+    sCharTexSize = 16.0f * ((f32)R_TEXT_CHAR_SCALE / 100.0f);
+    sCharTexScale = 1024.0f / ((f32)R_TEXT_CHAR_SCALE / 100.0f);
+
+    gSPTextureRectangle(gfx++, x << 2, y << 2, (x + sCharTexSize) << 2, (y + sCharTexSize) << 2, G_TX_RENDERTILE, 0, 0,
+                        sCharTexScale, sCharTexScale);
+
+    msgCtx->stateTimer++;
+
+    *p = gfx;
+}
+
+#define MESSAGE_SPACE_WIDTH 6
+// font widths
+f32 sFontWidths[144] = {
+    /*               */ 8.0f,
+    /* !             */ 8.0f,
+    /* "             */ 6.0f,
+    /* #             */ 9.0f,
+    /* $             */ 9.0f,
+    /* %             */ 14.0f,
+    /* &             */ 12.0f,
+    /* '             */ 3.0f,
```

Comment:
think // comments on the other side of the data would probably look better no? clang will align them to line up with each other
personally not a fan of the large gap created by the lower entries pushing everything else out